### PR TITLE
Deprecate some implicit Any constructors for Set, Dict and WeakKeyDict

### PR DESCRIPTION
**Pull request by fredrikekre**
I think it is better to be explicit about types here. Also, the methods are barely used in base which should indicate that they are not very useful. Number of uses in base (excluding tests):
 - `Set()`: ~5 times~ 4 times
 - `Dict()`: 8 times
 - `WeakKeyDict()`: 1 time

See also #24974 for deprecation of the implicit `Any` `Array` constructors.

### CODE REVIEW

File: base/array.jl [hunk 1]
```diff
@@ -2210,7 +2210,7 @@ end
 
 function union(vs...)
     ret = Vector{promote_eltype(vs...)}()
-    seen = Set()
+    seen = Set{Any}()
```
Comment (ararslan) [inline base/array.jl]: If I'm not mistaken, it seems like in this case we could do better than `Any`, e.g.
```julia
T = promote_eltype(vs...)
ret = Vector{T}()
seen = Set{T}()
...
```

File: base/set.jl [hunk 2]
```diff
@@ -312,7 +311,7 @@ julia> unique(x -> x^2, [1, -1, 3, -3, 4])
 """
 function unique(f::Callable, C)
     out = Vector{eltype(C)}()
-    seen = Set()
+    seen = Set{Any}()
```
Comment (ararslan) [inline base/set.jl]: Similarly here, it seems this could be `Set{eltype(c)}()`

File: base/array.jl (hunk 1 above)
Comment (rfourquet) [reply base/array.jl]: I did this change in a pending [PR](https://github.com/JuliaLang/julia/pull/23528/files#diff-d39084ec0fef2ad65e4b5fe0fcfaab8eR106), so if it gets accepted (I hope!) maybe no need to bother here

**Comment by rfourquet** [comment]
I don't have a strong opinion, but I find it's not a big deal to be implicit here and is not such a suprising behavior, and at the REPL it's somewhat nicer to just write `Dict()` and have it just work. I agree that the decision of keeping or deprecating those should match that of `Vector()`.

**Comment by StefanKarpinski** [comment]
I have to say that I kind of agree. We should make sure that we're consistent here, but if you ask for a `Dict` or a `Set` without an element type, I find it natural to get a collection that can hold anything. We already have much more verbose syntax for dict construction than other languages, this seems to make it worse since we have to write `Dict{Any,Any}()` where other languages write `{}`. At least `Dict()` is reasonably concise.

**Comment by StefanKarpinski** [comment]
x-ref: https://github.com/JuliaLang/julia/pull/24974

**Comment by StefanKarpinski** [comment]
The number of usages of untyped structures in Base is pretty misleading: Base is some of the most carefully typed generic code around. These are useful for quick and dirty interactive use, not writing generic library code.

**Comment by fredrikekre** [comment]
Fair enough.

**Pull request closed without merging**